File: src/main/java/ar/com/flow/bankaccount/transaction/Transaction.java
package ar.com.flow.bankaccount.transaction;

import ar.com.flow.money.Money;
import lombok.AllArgsConstructor;

import java.util.Collection;

@AllArgsConstructor
public class Transaction {
    private Collection<Precondition> preconditions;
    private final Collection<Step> steps;
    private final Money amount;

    public static Builder builder() {
        return new Builder();
    }

    public void execute() {
        preconditions.forEach(Precondition::check);

        var state = new State();

        for (Step step : steps) {
            try {
                step.execute(amount);
                state.completed(step, amount);
            } catch (Exception e) {
                state.rollback();
                return;
            }
        }
    }
}

File: src/main/kotlin/ar/com/flow/bankaccount/transaction/Transaction.kt
package ar.com.flow.bankaccount.transaction

import ar.com.flow.money.Money

class Transaction(
    private val preconditions: Collection<Precondition>,
    private val steps: List<Step>,
    private val amount: Money
) {

    fun execute() {
        preconditions.forEach{ precondition -> precondition.check() }

        val state = State()

        for (step in steps) {
            try {
                step.execute(amount)
                state.completed(step, amount)
            } catch (e: Exception) {
                state.rollback()
                return
            }
        }
    }
}